### PR TITLE
[PM-6702] 2# Region load in complete registration step

### PR DESCRIPTION
**Pull request by andrebispo5**
## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/PM-6702

## 📔 Objective

To complete the registration in the same environment where the user started, the region for a given email needs to be saved to state. This will allow the region to be loaded in the complete registration screen.

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed
  issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: app/src/test/java/com/x8bit/bitwarden/data/platform/repository/util/FakeEnvironmentRepository.kt [hunk 1]
```diff
@@ -10,6 +10,9 @@ import kotlinx.coroutines.flow.asStateFlow
  * A faked implementation of [EnvironmentRepository] based on in-memory caching.
  */
 class FakeEnvironmentRepository : EnvironmentRepository {
+    private var saveCurrentEnvironmentForEmailCalled = false
```
Comment (dseverns-livefront) [inline app/src/test/java/com/x8bit/bitwarden/data/platform/repository/util/FakeEnvironmentRepository.kt:13]: 💭  do we need these `var`s could we get rid of them and just have 

```kotlin
override fun saveCurrentEnvironmentForEmail(userEmail: String) = Unit
...
override fun loadEnvironmentForEmail(userEmail: String): Boolean  = true

Comment (andrebispo5) [reply app/src/test/java/com/x8bit/bitwarden/data/platform/repository/util/FakeEnvironmentRepository.kt:13]: Thanks you are right I added them but ended up not needing the inner state.

File: app/src/main/java/com/x8bit/bitwarden/data/platform/repository/EnvironmentRepositoryImpl.kt [hunk 2]
```diff
@@ -55,4 +56,19 @@ class EnvironmentRepositoryImpl(
...
+            ?: return false
+        environment = urls.toEnvironmentUrls()
+        return true
+    }
+
+    override fun saveCurrentEnvironmentForEmail(userEmail: String) =
+        environmentDiskSource
+            .storePreAuthEnvironmentUrlDataForEmail(
+                userEmail,
+                environment.environmentUrlData,
```
Comment (david-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/platform/repository/EnvironmentRepositoryImpl.kt:72]: Can we get some names on these params

File: app/src/test/java/com/x8bit/bitwarden/data/platform/datasource/disk/EnvironmentDiskSourceTest.kt [hunk 3]
```diff
@@ -60,6 +61,46 @@ class EnvironmentDiskSourceTest {
...
+        )
+        assertEquals(
+            json.encodeToJsonElement(mockUrls),
+            json.parseToJsonElement(requireNotNull(actual)),
+        )
+    }
+    companion object {
+        private const val EMAIL = "email@example.com"
+        private const val EMAIL_VERIFICATION_URLS_KEY = "bwPreferencesStorage:emailVerificationUrls"
+        private const val PRE_AUTH_URLS_KEY = "bwPreferencesStorage:preAuthEnvironmentUrls"
```
Comment (david-livefront) [inline app/src/test/java/com/x8bit/bitwarden/data/platform/datasource/disk/EnvironmentDiskSourceTest.kt:102]: You can just put these down below no need for the companion object

File: app/src/test/java/com/x8bit/bitwarden/data/platform/datasource/disk/EnvironmentDiskSourceTest.kt [hunk 4]
```diff
@@ -25,11 +28,9 @@ class EnvironmentDiskSourceTest {
 
     @Test
     fun `preAuthEnvironmentUrlData should pull from and update SharedPreferences`() {
-        val environmentKey = "bwPreferencesStorage:preAuthEnvironmentUrls"
-
         // Shared preferences and the repository start with the same value.
         assertNull(environmentDiskSource.preAuthEnvironmentUrlData)
-        assertNull(fakeSharedPreferences.getString(environmentKey, null))
+        assertNull(fakeSharedPreferences.getString(PRE_AUTH_URLS_KEY, null))
```
Comment (david-livefront) [inline app/src/test/java/com/x8bit/bitwarden/data/platform/datasource/disk/EnvironmentDiskSourceTest.kt:33]: Why change this?

Comment (andrebispo5) [reply app/src/test/java/com/x8bit/bitwarden/data/platform/datasource/disk/EnvironmentDiskSourceTest.kt:33]: Just for consistency and future reusability.

File: app/src/main/java/com/x8bit/bitwarden/data/platform/repository/EnvironmentRepositoryImpl.kt [hunk 5]
```diff
@@ -55,4 +56,19 @@ class EnvironmentRepositoryImpl(
             }
             .launchIn(scope)
     }
+
+    override fun loadEnvironmentForEmail(userEmail: String): Boolean {
+        val urls = environmentDiskSource
+            .getPreAuthEnvironmentUrlDataForEmail(userEmail)
+            ?: return false
+        environment = urls.toEnvironmentUrls()
```
[the remark this answers is not in the document]
Comment (andrebispo5) [reply app/src/main/java/com/x8bit/bitwarden/data/platform/repository/EnvironmentRepositoryImpl.kt:64]: It won't as explained in pm.

[the remark this answers is not in the document]
Comment (dseverns-livefront) [reply app/src/main/java/com/x8bit/bitwarden/data/platform/repository/EnvironmentRepositoryImpl.kt:64]: I believe I see the purpose of this, but I think this does get a little confusing. Feel free to correct me, the scenario where these would be called is we are entering the preAuth registration flow and need to update the environment for a "not yet active" account. We forcibly update the data source to return the environment data for this account when we read from the `environment` field. If the active account changes the environment will get updated via the `userStateFlow` collection as it does currently. I do think this is where I get uncomfortable as we are not mutating this value elsewhere outside of the flow.

💭 Would it be possible for us to just get back the specific environment from this call to `loadEnvironmentForEmail` directly?

**Review by david-livefront — APPROVED** [review]

**Pull request merged**